src/provider/controller/IntentHandler.ts

```diff
@@ -40,74 +40,76 @@ export class IntentHandler {
     public async raise(intent: Intent): Promise<IntentResolution> {
         if (hasTarget(intent)) {
             return this.raiseWithTarget(intent);
-        }
-
-        const apps: Application[] = await this._model.getApplicationsForIntent(intent.type);
-
-        if (apps.length === 0) {
-            throw new FDC3Error(ResolveError.NoAppsFound, 'No applications available to handle this intent');
-        } else if (apps.length === 1) {
-            // Resolve intent immediately
-            return this.fireIntent(intent, apps[0]);
         } else {
-            // Prompt the user to select an application to use
-            return this.queueResolve(intent);
+            return this.startResolve(intent);
         }
     }
 
     private async raiseWithTarget(intent: IntentWithTarget): Promise<IntentResolution> {
-        let appInfo: Application|null;
-
-        const appWindows = this._model.findWindowsByAppName(intent.target);
+        const apps = await this._model.getApplicationsForIntent(intent.type, intent.context.type);
+        const targetApp = apps.find(app => app.name === intent.target);
 
-        if (appWindows.length > 0) {
-            // Target app is running -> fire intent at it
-            appInfo = appWindows[0].appInfo;
+        if (targetApp !== undefined) {
+            // Target intent handles intent with given context, so fire
+            return this.fireIntent(intent, targetApp);
         } else {
-            // Target app not running -> Try to find in directory
-            appInfo = await this._directory.getAppByName(intent.target);
-            if (!appInfo) {
+            // Target intent does not handles intent with given, so determine why and throw an error
+            const targetInDirectory = await this._directory.getAppByName(intent.target);
+            const targetRunning = this._model.findWindowsByAppName(intent.target).length > 0;
+
+            if (!targetInDirectory && !targetRunning) {
                 throw new FDC3Error(
                     ResolveError.TargetAppNotAvailable,
                     `Couldn't resolve intent target '${intent.target}'. No matching app in directory or currently running.`
                 );
-            }
-
-            // Target app is in directory -> ensure that it handles intent
-            if (!(appInfo.intents || []).some(appIntent => appIntent.name === intent.type)) {
-                throw new FDC3Error(ResolveError.TargetAppDoesNotHandleIntent, `App '${intent.target}' does not handle intent '${intent.type}'`);
+            } else {
+                throw new FDC3Error(
+                    ResolveError.TargetAppDoesNotHandleIntent,
```

Comment:
Does this branch always correspond to the app not handling the intent? Are we sure there's no other case were we would fall into this section?

Reply:
Not that I can see. We know at this point it's either running or in the directory, so this seems like the remaining alternative. I have though changes how we test for if it's running, checking the environment, rather than for windows in the model, as that may give a misleading result if an app has windows that don't register with FDC3

Reply:
Hmmm, ok.